server/blob/blob-joining/src/test/java/org/apache/james/blob/joining/JoiningBlobStoreTest.java

```diff
@@ -0,0 +1,213 @@
+    void saveShouldWriteToPrimary() throws Exception {
+        BlobStore joiningBlobStore = testee();
+        BlobId blobId = joiningBlobStore.save(BLOB_CONTENT).get();
+
+        assertThat(PRIMARY_BLOB_STORE.readBytes(blobId).get())
+            .isEqualTo(BLOB_CONTENT);
+    }
+
+    @Test
+    void saveShouldNotWriteToSecondary() throws Exception {
+        BlobStore joiningBlobStore = testee();
+        BlobId blobId = joiningBlobStore.save(BLOB_CONTENT).get();
+
+        assertThat(SECONDARY_BLOB_STORE.readBytes(blobId).get())
+            .isEmpty();
+    }
+
+    @Test
+    void saveShouldNotWriteToSecondaryEvenWhenPrimaryFails() throws Exception {
+        JoiningBlobStore joiningBlobStore = new JoiningBlobStore(THROWING_BLOB_STORE, SECONDARY_BLOB_STORE);
+        assertThatThrownBy(() -> joiningBlobStore.save(BLOB_CONTENT_CAUSES_THROWING_DIRECTLY))
+            .isInstanceOf(RuntimeException.class);
+
+        assertThat(SECONDARY_BLOB_STORE.size())
+            .isEqualTo(0);
+    }
+
+    @Test
+    void saveShouldNotWriteToSecondaryEvenWhenPrimaryCompleteExceptionally() throws Exception {
+        JoiningBlobStore joiningBlobStore = new JoiningBlobStore(THROWING_BLOB_STORE, SECONDARY_BLOB_STORE);
+        assertThat(joiningBlobStore.save(BLOB_CONTENT))
+            .isCompletedExceptionally();
+
+        assertThat(SECONDARY_BLOB_STORE.size())
+            .isEqualTo(0);
+    }
+
+    @Test
+    void saveInputStreamShouldWriteToPrimary() throws Exception {
+        BlobStore joiningBlobStore = testee();
+        BlobId blobId = joiningBlobStore.save(new ByteArrayInputStream(BLOB_CONTENT)).get();
+
+        assertThat(PRIMARY_BLOB_STORE.readBytes(blobId).get())
+            .isEqualTo(BLOB_CONTENT);
+    }
+
+    @Test
+    void saveInputStreamShouldNotWriteToSecondary() throws Exception {
+        BlobStore joiningBlobStore = testee();
+        BlobId blobId = joiningBlobStore.save(new ByteArrayInputStream(BLOB_CONTENT)).get();
+
+        assertThat(SECONDARY_BLOB_STORE.readBytes(blobId).get())
+            .isEmpty();
+    }
+
+    @Test
+    void saveInputStreamShouldNotWriteToSecondaryEvenWhenPrimaryFails() throws Exception {
+        JoiningBlobStore joiningBlobStore = new JoiningBlobStore(THROWING_BLOB_STORE, SECONDARY_BLOB_STORE);
+        assertThatThrownBy(() -> joiningBlobStore.save(new ByteArrayInputStream(BLOB_CONTENT_CAUSES_THROWING_DIRECTLY)))
+            .isInstanceOf(RuntimeException.class);
+
+        assertThat(SECONDARY_BLOB_STORE.size())
+            .isEqualTo(0);
+    }
+
+    @Test
+    void saveInputStreamShouldNotWriteToSecondaryEvenWhenPrimaryCompleteExceptionally() throws Exception {
+        JoiningBlobStore joiningBlobStore = new JoiningBlobStore(THROWING_BLOB_STORE, SECONDARY_BLOB_STORE);
+        assertThat(joiningBlobStore.save(new ByteArrayInputStream(BLOB_CONTENT)))
+            .isCompletedExceptionally();
+
+        assertThat(SECONDARY_BLOB_STORE.size())
+            .isEqualTo(0);
+    }
+}
```

Comment:
IMO you are not testing the folowing cases:

 - **read** should use secondary blob store when primary blob store is failing
 - **read** should use secondary blob store when primary blob store is failing in a future

(The fallback behaviour upon read should not be altered in case of failure)

Reply:
will be in next commit